### PR TITLE
Link cli scan with asset and agent group.

### PR DESCRIPTION
**Pull request by elyousfi5**

### CODE REVIEW

File: src/ostorlab/runtimes/local/models/models.py [hunk 1]
```diff
@@ -572,6 +650,44 @@ class Asset(Base):
         sqlalchemy.Integer, sqlalchemy.ForeignKey("scan.id"), nullable=True
     )
 
+    @staticmethod
+    def create_from_assets_def(
+        scan: Scan, assets: Optional[List[base_asset.Asset]]
```
Comment (benyissa) [inline src/ostorlab/runtimes/local/models/models.py:655]: 1. I would use only scan_id  in the param.
2. can't we create an asset without scan ?

Comment (elyousfi5) [reply src/ostorlab/runtimes/local/models/models.py:655]: Yes we can, using just create

Comment (elyousfi5) [reply src/ostorlab/runtimes/local/models/models.py]: Applied

File: src/ostorlab/cli/scan/run/assets/android_apk.py [hunk 2]
```diff
@@ -48,10 +50,24 @@ def android_apk(
...
     logger.debug("scanning assets %s", [str(asset) for asset in assets])
     try:
-        runtime.scan(
+        created_scan = runtime.scan(
             title=ctx.obj["title"],
             agent_group_definition=ctx.obj["agent_group_definition"],
             assets=assets,
         )
+
+        if isinstance(ctx.obj["runtime"], runtime_local.LocalRuntime) is True:
```
Comment (3asm) [inline src/ostorlab/cli/scan/run/assets/android_apk.py:59]: As discussed, update runtime to add a method to perform the action and the other runtimes don't do anything in this case.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: src/ostorlab/runtimes/local/models/models.py [hunk 3]
```diff
@@ -412,18 +424,51 @@ def create(
...
+    @staticmethod
+    def to_bytes(
+        type: str, value: Optional[Union[int, float, str, bool, List, Dict]]
+    ) -> bytes:
+        """Convert the value to bytes."""
+        if type == "string":
+            return value.encode(encoding="utf-8")
+        elif type in ("boolean", "bool"):
+            return (1 if value is True else 0).to_bytes(1, byteorder="big")
+        elif type in ("number", "int", "float"):
```
Comment (deadly-panda) [inline src/ostorlab/runtimes/local/models/models.py:451]: We only support `number`

File: src/ostorlab/runtimes/local/models/models.py [hunk 4]
```diff
@@ -412,18 +424,51 @@ def create(
...
+            return struct.pack("d", value)
+        elif type in ("array", "object"):
```
Comment (deadly-panda) [inline src/ostorlab/runtimes/local/models/models.py:453]: am not sure about `object`?

File: src/ostorlab/runtimes/local/runtime.py [hunk 5]
```diff
@@ -769,3 +768,34 @@ def dump_vulnz(self, scan_id: int, dumper: dumpers.VulnzDumper):
...
+        with models.Database() as session:
+            agent_group_db = models.AgentGroup.create_from_agent_group_def(
+                agent_group_definition
+            )
+            scan.agent_group_id = agent_group_db.id
+            session.add(scan)
+            session.commit()
+
+    def link_assets_scan(
+        self, scan_id: int, assets: Optional[List[base_asset.Asset]]
```
Comment (deadly-panda) [inline src/ostorlab/runtimes/local/runtime.py:792]:
```suggestion
        self, scan_id: int, assets: Optional[List[base_asset.Asset]] = None
```

File: src/ostorlab/runtimes/local/models/models.py [hunk 6]
```diff
@@ -472,6 +517,39 @@ def create(name: str, description: str, agents: Any) -> "AgentGroup":
             session.commit()
             return agent_group
 
+    @staticmethod
+    def create_from_agent_group_def(
```
Comment (3asm) [inline src/ostorlab/runtimes/local/models/models.py:521]:
```suggestion
    def create_from_agent_group_definition(
```

File: src/ostorlab/runtimes/local/models/models.py [hunk 7]
```diff
@@ -572,6 +650,44 @@ class Asset(Base):
         sqlalchemy.Integer, sqlalchemy.ForeignKey("scan.id"), nullable=True
     )
 
+    @staticmethod
+    def create_from_assets_def(
```
Comment (3asm) [inline src/ostorlab/runtimes/local/models/models.py:654]:
```suggestion
    def create_from_assets_definition(
```

File: src/ostorlab/runtimes/runtime.py [hunk 8]
```diff
@@ -96,3 +96,25 @@ def dump_vulnz(self, scan_id: int, dumper: dumpers.VulnzDumper):
         None
         """
         raise NotImplementedError()
+
+    def link_agent_group_scan(
```
Comment (3asm) [inline src/ostorlab/runtimes/runtime.py:100]: Should be marked as abstract method.

**Review by 3asm — CHANGES_REQUESTED** [review]

File: tests/conftest.py [hunk 9]
```diff
@@ -1100,3 +1106,81 @@ def ios_store(mocker: plugin.MockerFixture, db_engine_path: str) -> models.IosSt
...
+
+    mocker.patch(
+        "ostorlab.runtimes.local.runtime.LocalRuntime.can_run", return_value=True
+    )
+    mocker.patch("ostorlab.runtimes.local.runtime.LocalRuntime._create_network")
+    mocker.patch("ostorlab.runtimes.local.runtime.LocalRuntime._start_services")
+    mocker.patch("ostorlab.runtimes.local.runtime.LocalRuntime._start_pre_agents")
+    mocker.patch("ostorlab.runtimes.local.runtime.LocalRuntime._start_agents")
+    mocker.patch("ostorlab.runtimes.local.runtime.LocalRuntime._start_post_agents")
+    mocker.patch("ostorlab.runtimes.local.runtime.LocalRuntime._check_services_healthy")
```
Comment (deadly-panda) [inline tests/conftest.py:1181]: Can we avoid mocking private methods? or is it too much ?

File: src/ostorlab/runtimes/local/models/models.py (hunk 3 above)
Comment (elyousfi5) [reply src/ostorlab/runtimes/local/models/models.py:451]: in the schema yes, but in some agents we use others. Example nmap:
image

File: src/ostorlab/runtimes/local/models/models.py (hunk 4 above)
Comment (elyousfi5) [reply src/ostorlab/runtimes/local/models/models.py:453]: what about it ? 
image

File: tests/conftest.py (hunk 9 above)
Comment (elyousfi5) [reply tests/conftest.py:1181]: I tried, there is no way to get over it !

File: src/ostorlab/runtimes/local/models/models.py [hunk 10]
```diff
@@ -412,18 +424,51 @@ def create(
...
 
+    @staticmethod
+    def to_bytes(
+        type: str, value: Optional[Union[int, float, str, bool, List, Dict]]
+    ) -> bytes:
+        """Convert the value to bytes."""
+        if type == "string":
+            return value.encode(encoding="utf-8")
+        elif type in ("boolean", "bool"):
+            return (1 if value is True else 0).to_bytes(1, byteorder="big")
```
Comment (mohsinenar) [inline src/ostorlab/runtimes/local/models/models.py:450]: Why not bytes(value)??

File: src/ostorlab/runtimes/local/models/models.py [hunk 11]
```diff
@@ -572,6 +650,44 @@ class Asset(Base):
...
+
+        Args:
+            scan: The scan object.
+            assets: The list of assets to create.
+        """
+        networks: List[str] = []
+        links: List[str] = []
+        for asset in assets:
+            if (
+                isinstance(asset, ip.IP)
```
Comment (mohsinenar) [inline src/ostorlab/runtimes/local/models/models.py:667]: Explicit ?

**Review by mohsinenar — DISMISSED** [review]

File: src/ostorlab/runtimes/local/models/models.py (hunk 10 above)
Comment (elyousfi5) [reply src/ostorlab/runtimes/local/models/models.py:450]: im using this method to make it easy to get value from bytes

File: src/ostorlab/runtimes/local/models/models.py (hunk 3 above)
Comment (deadly-panda) [reply src/ostorlab/runtimes/local/models/models.py:451]: We do not force the type check when schema validation? how it was not spotted ..
The agent & the schema should be aligned..

Comment (3asm) [reply src/ostorlab/runtimes/local/models/models.py:451]: +1

**Review by 3asm — DISMISSED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**